pkg/controller/cluster/actuator.go

```diff
@@ -0,0 +1,34 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	clusterv1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+)
+
+// Actuator controls clusters on a specific infrastructure. All
+// methods should be idempotent unless otherwise specified.
+type Actuator interface {
+	// Create the cluster
```

Comment:
This interface made sense for machines since there is one object that was "the machine" (eg. does the literal VM exist). Does this interface really make sense for clusters? Perhaps a Reconcile,Delete interface makes more sense?

Reply:
Yea I think it would make sense to remove the exists method. Would you imagine the reconcile method checking the status field on the cluster to see if the cluster has already been created? Since that isn't really cloud provider specific, the cloud controller could check for the existence of the status field and call either create or update based on what it finds. Then we'd have create/update/delete instead of reconcile/delete. What do you think?

Reply:
I lean more towards the reconcile/delete pattern for this case since (in my mind) the cluster is actually backed by a bunch of disparate objects on the cloud provider (eg. some service accounts, perhaps a load balancer). Even if the controller had previously created everything, some things could have since been deleted. This can mean the state can be in a partially created/partially out of date state. If a cluster has some pieces missing and some pieces in need of updating, is that a create or an update? I am guessing it is a create the missing bits then update the existing ones. At that point, the pattern looks awfully a lot like a reconcile. Thoughts?

Reply:
That's a good point, I hadn't thought of the case where something external deleted the cloud resources and they need to be recreated. I'll update this to have a reconcile method instead of create/update/exists.

Another thing I would like your thoughts on, do you think the reconcile method should take a machine parameter too, instead of just a cluster? Then the cluster controller can look for the machine with that role and pass it into the reconcile method, similar to how the machine controller finds the cluster for the machines. If the cluster controller is going to be responsible for creating the master machine it seems like the flow needs to be: 
1. Master machine object is created 
2. Cluster object is created 
3. Other worker machines are created

Reply:
My 2-cents - long term, I believe it will be the cluster controller (or some equivalent) that will actually manage the control plane. There will be a cluster-level spec for the control plane and the said controller will make all the needed objects for it to be a reality (including master machine object(s)). 

Whether you want the cluster actuator to find the master machine or have the cluster controller pass it in could go either way in my mind.

Reply:
Ok, thanks! I agree with you that it makes sense for the cluster controller to create the machine object, which will make the code cleaner because the machine controller can still take care of creating that machine. This will involve some tweaking to the current configuration for clusters. I'll try to think through a way around this for now and have the reconcile method only take a cluster parameter because I think thats how we'll want it long term.